sdks/java/io/csv/src/main/java/org/apache/beam/sdk/io/csv/CsvIO.java

```diff
@@ -0,0 +1,384 @@
+
+      if (getTempDirectory() != null) {
+        builder.add(DisplayData.item("tempDirectory", getTempDirectory().getFilename()));
+      }
+    }
+
+    @Override
+    public PDone expand(PCollection<T> input) {
+      if (!input.hasSchema()) {
+        throw new IllegalArgumentException(
+            String.format(
+                "%s requires an input Schema. Note that only Row or user classes are supported. Consider using TextIO or FileIO directly when writing primitive types",
+                Write.class.getName()));
+      }
+
+      Schema schema = input.getSchema();
+      SerializableFunction<T, Row> toRowFn = input.getToRowFunction();
+      PCollection<Row> rows =
+          input.apply(MapElements.into(rows()).via(toRowFn)).setRowSchema(schema);
+      rows.apply("writeRowsToCsv", buildFileIOWrite().via(buildSink(schema)));
+      return PDone.in(input.getPipeline());
+    }
+
+    /**
+     * Builds a header using {@link CSVFormat} based on either a {@link Schema#sorted()} {@link
+     * Schema#getFieldNames()} if {@link #getSchemaFields()} is null or {@link #getSchemaFields()}.
+     */
+    String buildHeader(Schema schema) {
+      if (getSchemaFields() != null) {
+        return CsvUtils.buildHeaderFrom(getSchemaFields(), getCSVFormat());
+      }
+      return CsvUtils.buildHeaderFrom(schema.sorted(), getCSVFormat());
+    }
+
+    /** Builds a {@link Sink} for writing {@link Row} serialized using {@link CSVFormat}. */
+    Sink<Row> buildSink(Schema schema) {
+      List<String> schemaFields = null;
+      String header = null;
+      if (getSchemaFields() != null) {
+        schemaFields = getSchemaFields();
+      }
+      return Sink.<Row>builder()
+          .setPreamble(getPreamble())
+          .setHeader(buildHeader(schema))
+          .setFormatFunction(
+              CsvUtils.getRowToCsvStringFunction(schema, getCSVFormat(), schemaFields))
+          .build();
+    }
+
+    /** Builds a {@link FileIO.Write} with a {@link Sink}. */
+    FileIO.Write<Void, Row> buildFileIOWrite() {
+      checkArgument(getFilenamePrefix() != null, "to() is required");
+
+      ResourceId prefix =
+          FileSystems.matchNewResource(getFilenamePrefix(), false /* isDirectory */);
+
+      FileIO.Write<Void, Row> write =
+          FileIO.<Row>write()
+              .to(prefix.getCurrentDirectory().toString())
+              .withPrefix(Objects.requireNonNull(prefix.getFilename()));
+
+      if (getCompression() != null) {
+        write = write.withCompression(getCompression());
+      }
+
+      if (getNumShards() != null) {
+        write = write.withNumShards(getNumShards());
+      }
+
+      if (getFilenameSuffix() != null) {
+        write = write.withSuffix(getFilenameSuffix());
+      }
+
+      if (getTempDirectory() != null) {
+        write = write.withTempDirectory(Objects.requireNonNull(getTempDirectory().getFilename()));
+      }
+
+      return write;
+    }
```

Comment:
nit: Consider inlining these into expand since these are small methods and we don't expect them to be used outside of expand()